sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -503,6 +503,18 @@ trait HiveTypeCoercion {
       // Hive lets you do aggregation of timestamps... for some reason
       case Sum(e @ TimestampType()) => Sum(Cast(e, DoubleType))
       case Average(e @ TimestampType()) => Average(Cast(e, DoubleType))
+
+      case Coalesce(es) if es.map(_.dataType).distinct.size > 1 =>
+        val dt: Option[DataType] = Some(NullType)
+        val rt = es.map(_.dataType).foldLeft(dt)((r, c) => if (r != None) {
+          findTightestCommonType(r.get, c)
+        } else {
+          None
+        })
+        rt match {
+          case Some(ddt) => Coalesce(es.map(Cast(_, ddt)))
+          case None => sys.error("Cannot do type coercion over coalesce")
```

Comment:
Seems it will be good to also let users know the data types of those expressions in the `Coalesce`. Also, maybe `type coercion` is not very user friendly. How about "Could not determine the return type of Coalesce" and then provide the types of expressions?